### PR TITLE
added copy/download emails feature for instructor user in student and graders page

### LINKED ISSUE #1662: Copy-paste students/graders email
For an instructor user, from the 'students' tab, add a download/copy-paste button for all student email addresses.  Options for 'all registered (non null)', 'all students in section X' (for each section), and 'all users (including null)'.  Also would be good to have 'all graders (full, limited, & instructor)', 'all limited access', and 'all full access graders'.

### PR DESCRIPTION
**Pull request by tushargr**
Closes #1662 
Modified 3 files:
1. `UsersView.php` - made two functions consisting of html code for copy/download form. One function consist of html code for copy/download form for students tab and other function for graders tab. Both form have copy as well as download option.
2. `server.js` - made 4 functions. 
One function for changing visibility of rendered html form. second function to do copy / download on clicking button on form.
These two functions are implemented both for students and graders. 
3. `UsersController.php` - made two functions that will render html forms created in `UsersView.php`

### CODE REVIEW

**Comment by tushargr** [comment]
@bmcutler can you check why some checks aren't successful

File: site/app/controllers/admin/UsersController.php [hunk 1]
```diff
@@ -69,6 +71,16 @@ private function renderUserForm($action, $use_database) {
         $this->core->getOutput()->renderOutput(array('admin', 'Users'), 'userForm', $reg_sections, $rot_sections, $action, $use_database);
     }
 
+    private function renderCopyStudentEmailForm($students, $use_database) {
```
Comment (tushargr) [inline site/app/controllers/admin/UsersController.php]: This function renders form that copy/download emails for student tab

File: site/app/controllers/admin/UsersController.php [hunk 2]
```diff
@@ -69,6 +71,16 @@ private function renderUserForm($action, $use_database) {
...
+        $reg_sections = $this->core->getQueries()->getRegistrationSections();
+        $this->core->getOutput()->renderOutput(array('admin', 'Users'), 'copyStudentEmailForm',$students, $reg_sections, $use_database);
+    }
+
+
+    private function renderCopyGraderEmailForm($graders, $use_database) {
```
Comment (tushargr) [inline site/app/controllers/admin/UsersController.php]: This function renders form that copy/download emails for graders tab

File: site/app/views/admin/UsersView.php [hunk 3]
```diff
@@ -295,6 +297,133 @@ public function userForm($reg_sections, $rot_sections, $action, $use_database=fa
         return $return;
     }
 
+    /**
+     * Creates the form box to be displayed when copying or downloading emails on the students pages
+     * @param User[] $students  
+     * @param array  $reg_sections associative array representing registration sections in the system
+     * @param bool   $use_database
+     * @return string
+     */
```
Comment (tushargr) [inline site/app/views/admin/UsersView.php]: this function consist of html code for copy/download form that will be rendered for student tab

File: site/app/views/admin/UsersView.php [hunk 4]
```diff
@@ -295,6 +297,133 @@ public function userForm($reg_sections, $rot_sections, $action, $use_database=fa
...
+        return $return;
+    }
+
+
+/**
+     * Creates the form box to be displayed when copying or downloading emails on the graders pages
+     * @param User[] $graders
+     * @param bool   $use_database
+     * @return string
+     */
```
Comment (tushargr) [inline site/app/views/admin/UsersView.php]: this function consist of html code for copy/download form that will be rendered for graders tab

File: site/public/js/server.js [hunk 5]
```diff
@@ -115,6 +135,106 @@ function newClassListForm() {
     $('[name="upload"]', form).val(null);
 }
 
+function copyStudentEmail(code,emails,emails_corresponding_section) {
```
Comment (tushargr) [inline site/public/js/server.js]: following two functions are js functions which actually performs copy or download when button on form is clicked.
1st function is for student tab and next for grader tab for instructor user

**Comment by tushargr** [comment]
@bmcutler  following is the video link for my deployment having the copy/download feature-
https://drive.google.com/file/d/1B6Vc4fBuvc_Bl9pj-Ob6Ewck4it0YBSx/view?usp=sharing

**Comment by tushargr** [comment]
Someone review the pull request @bmcutler @MasterOdin

**Comment by bmcutler** [comment]
Hi Tushar,

Nice work.  Again, first I'm going to focus on UI stuff.  Code comments later.

When we copy the emails to the clipboard, I think we want separators (probably a comma?  a quick google search says that's the standard except for outlook).

Rather than a pulldown menu or radio buttons, I think we should have checkboxes for each registration section, and by default have all sections selected.  There should be a checkbox for NULL section students.  By default the NULL section should be unselected (usually you do not want to email students who have dropped).  

Underneath the registration sections, let's have checkboxes for instructors, full-access graders and limited access graders.  By default, all of these selected as well.  The typical use case for this will be to send an announcement to the whole class, and the instructor wants the TAs to get the announcement as well.

Note: We'll need to ensure that instructors/graders who are assigned to a registration section (usually they are "null" because they aren't students in the course, but not always) are handled correctly.   For example: A user who is both a limited access grader and in registration section 2 should be on the list if either of those checkboxes is selected.

For the download file, rather than a text file, it would be more helpful to have a .csv that has preferred first name, last name, username, email, registration section, rotating section, and user group as columns.  Then the button from the Students/Graders tab to launch this popup should be called "Download users" or "Download graders" and then from the popup label the buttons "Download .csv" or "Copy email addresses to clipboard".

Thanks
Barb

**Comment by tushargr** [comment]
Sorry I saw your reply late....i will start working on this UI change

**Comment by tushargr** [comment]
Hi, @bmcutler  Made the UI changes as sugggested. Also made CSV download feature.
link for my deployment- https://drive.google.com/file/d/1VQ0G97P41sGQWlUxwnvyiOcQvGIaolMW/view?usp=sharing

**Comment by tushargr** [comment]
@bmcutler @MasterOdin  @tcyrus review the pull request

**Comment by tushargr** [comment]
@bmcutler  does the pull request require any changes to be made?

**Comment by bmcutler** [comment]
@tushargr I've sent by email a patch with minor wording changes.

@emaicus or @MasterOdin can you check over the code quickly and confirm that there are no security issues?  Is it possible for a non-instructor to download the classlist by URL hacking or other methods?
Thanks

**Pull request merged**